libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkVectorLibrary.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.nativeaccess.jdk;
+
+import org.elasticsearch.nativeaccess.lib.VectorLibrary;
+
+import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.MemorySegment;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+
+import static java.lang.foreign.ValueLayout.ADDRESS;
+import static java.lang.foreign.ValueLayout.JAVA_BYTE;
+import static java.lang.foreign.ValueLayout.JAVA_INT;
+import static org.elasticsearch.nativeaccess.jdk.LinkerHelper.downcallHandle;
+
+public final class JdkVectorLibrary implements VectorLibrary {
+
+    static {
+        loadLibrary();
+    }
+
+    @SuppressWarnings("removal")
+    private static void loadLibrary() {
+        AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+            System.loadLibrary("vec");
+            return null;
+        });
+    }
+
+    public JdkVectorLibrary() {}
+
+    private static final MethodHandle stride$mh = downcallHandle("stride", FunctionDescriptor.of(JAVA_INT));
+    private static final MethodHandle dot8s$mh = downcallHandle("dot8s", FunctionDescriptor.of(JAVA_INT, ADDRESS, ADDRESS, JAVA_INT));
+
+    // Stride of the native implementation - consumes this number of bytes per loop invocation.
+    // There must be at least this number of bytes/elements available when going native
+    static final int STRIDE = 32;
+
+    static {
+        assert STRIDE > 0 && (STRIDE & (STRIDE - 1)) == 0 : "Not a power of two";
+        assert stride() == STRIDE;
+    }
+
+    /**
+     * Computes the dot product of given byte vectors.
+     * @param a address of the first vector
+     * @param b address of the second vector
+     * @param length the vector dimensions
+     */
+    // @Override
+    static int dotProduct(MemorySegment a, MemorySegment b, int length) {
+        assert length >= 0;
+        if (a.byteSize() != b.byteSize()) {
+            throw new IllegalArgumentException("dimensions differ: " + a.byteSize() + "!=" + b.byteSize());
+        }
+        if (length > a.byteSize()) {
+            throw new IllegalArgumentException("length: " + length + ", greater than vector dimensions: " + a.byteSize());
+        }
+        int i = 0;
+        int res = 0;
+        if (length >= STRIDE) {
+            i += length & ~(STRIDE - 1);
+            res = dot8s(a, b, i);
+        }
+
+        // tail
+        for (; i < length; i++) {
```

Comment:
Just for the record, as we discussed this offline: would it make sense to avoid processing the tail here and have a STRIDE at all, and just use native but scalar code to perform the dotproduct on the tail?
This will also leave more freedom to different implementations to optimize (e.g. different strides on different architectures, or auto-vectorization/more efficient compilation of the tail, etc.)

Reply:
I'm currently undecided about this. For now, let's leave it as it. We can revisit later.